### PR TITLE
Convert DeviceComm into an interface and add handle exchange

### PR DESCRIPTION
**Pull request by DHowett**
This commit replaces DeviceComm with the interface IDeviceComm and the
concrete implementation type ConDrvDeviceComm. This work is done in
preparation for different device backends.

In addition to separating out ConDrv-specific behavior, I've introduced
a "handle exchange" interface.

HANDLE EXCHANGE
---------------

There are points where we give ConDrv opaque handle identifiers to our
input buffer, output buffer and process data. The exact content of the
opaque identifier is meaningless to ConDrv: the driver's only
interaction with these identifiers is to hold onto them and send back
whichever are pertinent for each API call.

Because of that, we used the raw register-width pointer value of the
input buffer, output buffer or process data _as_ the opaque handle
value.

This works very well for ConDrv <-> conhost using the ConDrvDeviceComm.
It works less well for something like the "logging" DeviceComm that will
log packets to a file: those packets *cannot* contain pointer values (!)

To address this, and to afford flexibility to DeviceComm implementers,
I've introduced a two-member complement of handle management functions:

* `ULONG_PTR PutHandle(void*)` registers an object with the DeviceComm
  and returns an opaque identifier.
* `void* GetHandle(ULONG_PTR)` takes an opaque identifier and returns
  the original object.

ConDrvDeviceComm implements PutHandle and GetHandle by casting the
object pointer to the "opaque handle value", which maintains wire format
compatibility[1] with revisions of conhost prior to this change.

Simpler DeviceComm implementations that require handle tracking but
cannot bear raw pointers can implement these functions by returning an
autoincrementing ID (or similar) and registering the raw object pointer
in a mapping.

I've audited all existing handle exchanges with the driver and updated
them to use Put/GetHandle.

(I intended to add DestroyHandle, but we are not equipped for handle
removal at the moment. ConsoleHandleData/ConsoleProcessHandle are
destroyed during wait routine completion, on client disconnect, etc.
This does mean that an id<->pointer mapping will grow without bound, but
at a cost of ~8 bytes per entry and a short-lived console session I am
not too concerned about the cost.)

[1] Wire format compatibility is not required, and later we may want to
switch ConDrvDeviceComm to `EncodePointer` and `DecodePointer` just to
insulate us against a spurious ConDrv packet allowing for an arbitrary
4/8-byte read and subsequent liftoff into space.

### CODE REVIEW

**Comment by KalleOlaviNiemitalo** [comment]
> those packets _cannot_ contain pointer values

Is that because pointers in a log would thwart ASLR?

> at a cost of ~8 bytes per entry and a short-lived console session I am not too concerned about the cost

If conhost ever started tracking handles by default, that might be harmful for services that repeatedly start child processes. IIS and its w3wp.exe worker processes come to mind; w3wp.exe uses the console subsystem, but I don't know whether its parent process lets it inherit a console.

> later we may want to switch ConDrvDeviceComm to `EncodePointer` and `DecodePointer`

A related post about the security of those functions: [The history of the EncodePointer function for obfuscating pointers | The Old New Thing](https://devblogs.microsoft.com/oldnewthing/20201113-00/?p=104447)

**Comment by DHowett** [comment]
> thwart ASLR

Oh no, it’s because I want to capture binary ConDrv packet logs for replay and needed a stable identifier from session to session :smile:

Having conhost keep track of all handles isn’t on my radar, but that is an excellent call-out. Thanks!

Raymond’s blog is the only reason I even knew to suggest those functions :smile:

**Comment by DHowett** [comment]
> Binary logs

This was pulled out of branch `dev/duhowett/hax/conhost_dump_replay` which for the most part works to implement logging. Running a conhost build from that branch produces a single log (`conhost_log.bin`) in PWD and running it with that log as an argument replays it. My plan is to have reproduction of bugs isolatable from client applications by letting us keep playing back raw packet captures. It’s also able to replay a log over ConPTY, though that is not its purpose.

**Comment by miniksa** [comment]
> Wire format compatibility
LOL compatibility with giving out arbitrary heap addresses.

File: src/server/ConDrvDeviceComm.cpp [hunk 1]
```diff
@@ -150,3 +150,15 @@ DeviceComm::~DeviceComm()
 
     return S_OK;
 }
+
+[[nodiscard]] ULONG_PTR ConDrvDeviceComm::PutHandle(const void* handle)
```
Comment (miniksa) [inline src/server/ConDrvDeviceComm.cpp:154]: nit: doc comments, technically, as the rest of this file has 'em.

**Review by miniksa — APPROVED** [review]

File: src/server/ConDrvDeviceComm.cpp (hunk 1 above)
Comment (DHowett) [reply src/server/ConDrvDeviceComm.cpp:154]: reckon i oughtta

**Comment by DHowett** [comment]
@zadjii-msft (since you're assigned) i cleared the Draft status on this one

**Comment by zadjii-msft** [comment]
Okay, I'll do a review pass on PRs probably tomorrow.

File: src/server/ProcessHandle.cpp [hunk 2]
```diff
@@ -35,12 +35,12 @@ ConsoleProcessHandle::ConsoleProcessHandle(const DWORD dwProcessId,
     }
 }
 
-CD_CONNECTION_INFORMATION ConsoleProcessHandle::GetConnectionInformation() const
+CD_CONNECTION_INFORMATION ConsoleProcessHandle::GetConnectionInformation(IDeviceComm* deviceComm) const
```
Comment (zadjii-msft) [inline src/server/ProcessHandle.cpp:38]: This might warrant a doc comment now too

**Review by zadjii-msft — APPROVED** [review]
I'm okay with this but pls add those doc comments

File: src/server/ConDrvDeviceComm.cpp (hunk 1 above)
Comment (DHowett) [reply src/server/ConDrvDeviceComm.cpp:154]: Docced.

File: src/server/ProcessHandle.cpp (hunk 2 above)
Comment (DHowett) [reply src/server/ProcessHandle.cpp:38]: Docced!

**Pull request merged**